emg_publisher:
  ros__parameters:
    # rfcomm port Shimmer MAC address is bound to
    rfcomm_port_num: 0
    # Chip gain
    gain: 12
    # Internal chip data rate. Must be > than sampling rate
    emg_data_rate: 125 # Hz
    # Shimmer sampling rate
    sampling_rate: 992.96969699 # Hz
    # Stream 1 Hz square wave with +-1 amplitude test signal instead of acquired
    stream_test_signal: False
    # Zero ADC offset using test signal before data acquisition
    zero_test_signal: 4 # Calibration duration in seconds (0 - no calibration)
    # Size of the queue for asynchronous publishing on topic
    queue_size: 10

# TODO: ROS2 cannot parse nested lists, make flat, and reshape in code.
emg_processor:
  ros__parameters:
    # Window size for fft filtering (as s fraction of sampling_rate)
    window_size_fft_relative: 0.5
    # Window size for smoothing (sec) - recommended between 0.05 and 0.5 sec
    # Optimal window size - 302 data points
    window_size_ema_relative: 0.304138
    # Band-stop frequencies (Hz) - a list of low, high ranges [low, high, low, high ...]
    band_stop: []
    # High pass filter cutoff frequency (Hz), remove DC component, use negative for None
    hpf_cutoff: 0.1
    # Low pass filter cutoff frequency (Hz), remove high frequency noise, use large 
    # value for None
    lpf_cutoff: 203.0
    # Custom FFT mask multipliers in frequency domain (amplitudes multiplied with factor)
    # [low, high, multiplier, low, high, multiplier, ...]
    fft_mask: [
      1.9, 2.1, 0.05, 3.9, 4.1, 0.1, 5.9, 6.1, 0.2, 7.9, 8.1, 0.25,
      9.9, 10.1, 0.5, 11.9, 12.1, 0.5, 13.9, 14.1, 0.5, 15.9, 16.1, 0.75, 
      17.9, 18.1, 1.0, 19.9, 20.1, 1.25, 21.9, 22.1, 1.75, 23.9, 24.1, 1.5,
      25.9, 26.1, 1.75, 27.9, 28.1, 2.0, 29.9, 30.1, 2.0, 31.9, 32.1, 2.5,
      33.9, 34.1, 2.5, 35.9, 36.1, 2.5, 37.9, 38.1, 2.5, 39.9, 40.1, 2.5,
      41.9, 42.1, 2.25, 43.9, 44.1, 2.25, 45.9, 46.1, 1.5, 47.9, 48.1, 1.25,
      49.9, 50.1, 0.375, 51.9, 52.1, 1.5, 53.9, 54.1, 2.25, 55.9, 56.1, 1.75,
      57.9, 58.1, 2.25, 59.9, 60.1, 2.25, 61.9, 62.1, 3.5, 63.9, 64.1, 2.75,
      65.9, 66.1, 3.25, 67.9, 68.1, 3.25, 69.9, 70.1, 2.5, 71.9, 72.1, 3.25, 
      73.9, 74.1, 4.25, 75.9, 76.1, 3.25, 77.9, 78.1, 3.25, 79.9, 80.1, 2.75,
      81.9, 82.1, 3.75, 83.9, 84.1, 3.5, 85.9, 86.1, 3.5, 87.9, 88.1, 3.5,
      89.9, 90.1, 4.75, 91.9, 92.1, 4.0, 93.9, 94.1, 5.0, 95.9, 96.1, 4.5,
      97.9, 98.1, 3.5, 99.9, 100.1, 4.25, 101.9, 102.1, 4.25, 103.9, 104.1, 4.5,
      105.9, 106.1, 4.5, 107.9, 108.1, 4.25, 109.9, 110.1, 5.5, 111.9, 112.1, 5.5, 
      113.9, 114.1, 5.5, 115.9, 116.1, 5.5, 117.9, 118.1, 5.5, 119.9, 120.1, 5.5,
      121.9, 122.1, 4.5, 123.9, 202.1, 5.25
      ]
    # Rectify EMG signal after filtering (absolute value): true/false
    rectify: True
    # Smooth signal using weighted exponential moving average: true/false
    smooth: True
    # Exponential moving average decreasing factor (0 - 1),
    # 0 - simple moving average
    # Recommendation to use values < 0.01 (older values will have too low weights 
    # otherwise) 
    ema_decay: 2.5e-4
    # Queue size for FFT and smoothing
    queue_size: 10

emg_calibration:
  ros__parameters:
    experiment_duration: 10.0  # seconds
    # Sampling rate has to be read from emg_publisher node
    service_name: '/emg_publisher/get_parameters'
